File: limitless/app/models/message.py
from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING
from uuid import UUID, uuid4

from sqlalchemy import DateTime, ForeignKey, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base

if TYPE_CHECKING:
    from app.models.entity import Entity
    from app.models.message_embedding import MessageEmbedding
    from app.models.tenant import Tenant
    from app.models.user import User


class Message(Base):
    __tablename__ = "messages"

    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    tenant_id: Mapped[UUID] = mapped_column(ForeignKey("tenants.id"), nullable=False, index=True)
    user_id: Mapped[UUID] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    event_time: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)
    ingested_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, nullable=False)

    tenant: Mapped[Tenant] = relationship("Tenant", back_populates="messages")
    user: Mapped[User] = relationship("User", back_populates="messages")
    embedding: Mapped[MessageEmbedding | None] = relationship(
        "MessageEmbedding", back_populates="message", uselist=False
    )
    entities: Mapped[list[Entity]] = relationship("Entity", back_populates="message")

File: limitless/frontend/src/models/Message.ts
export interface Message {
  id: string;
  tenantId: string;
  userId: string;
  content: string;
  eventTime: string;
  ingestedAt: string;
}

export function toMessage(raw: {
  id: string;
  tenant_id: string;
  user_id: string;
  content: string;
  event_time: string;
  ingested_at: string;
}): Message {
  return {
    id: raw.id,
    tenantId: raw.tenant_id,
    userId: raw.user_id,
    content: raw.content,
    eventTime: raw.event_time,
    ingestedAt: raw.ingested_at,
  };
}
